### PR TITLE
ADEN-5787 Fallback to bid offer when there is no ad for Porvata

### PR DESCRIPTION
**Pull request by fraszczakszymon**

### CODE REVIEW

File: extensions/wikia/AdEngine/js/wrappers/prebid.js [hunk 1]
```diff
@@ -31,6 +31,28 @@ define('ext.wikia.adEngine.wrappers.prebid', [
...
+		var bids,
+			videoBid = null;
+
+		if (!win.pbjs) {
+			return null;
+		}
+
+		bids = win.pbjs.getBidResponsesForAdUnitCode(slotName).bids || [];
+
+		bids.forEach(function (bid) {
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/wrappers/prebid.js]: you could use `filter` and `reduce` ;)

File: extensions/wikia/AdEngine/js/video/player/porvata/googleImaPlayerFactory.js [hunk 2]
```diff
@@ -164,7 +166,16 @@ define('ext.wikia.adEngine.video.player.porvata.googleImaPlayerFactory', [
 			false
 		);
 
+		adsLoader.addEventListener(win.google.ima.AdErrorEvent.Type.AD_ERROR, function (event) {
+			var emptyVastErrorCode = win.google.ima.AdError.ErrorCode.VAST_EMPTY_RESPONSE;
+
+			if (event.getError && event.getError().getErrorCode() === emptyVastErrorCode) {
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/video/player/porvata/googleImaPlayerFactory.js]: maybe we can check type of `getError` before calling it?

File: extensions/wikia/AdEngine/js/template/porvata.js [hunk 3]
```diff
@@ -126,6 +131,22 @@ define('ext.wikia.adEngine.template.porvata', [
 			}
 		});
 
+		if (params.slotNameForFallbackBid) {
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/template/porvata.js]: Please consider using boolean flag for enabling fallbackBid and using `params.slotName`

File: extensions/wikia/AdEngine/js/template/porvata.js [hunk 4]
```diff
@@ -126,6 +131,22 @@ define('ext.wikia.adEngine.template.porvata', [
...
+			video.addEventListener('wikiaEmptyAd', function () {
+				var fallbackBid = prebid.getWinningVideoBidBySlotName(params.slotNameForFallbackBid, fallbackBidders);
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/template/porvata.js]: can we avoid calling this method when we know we won't use this bid? (`fallbackAdRequested = true`)

File: extensions/wikia/AdEngine/js/wrappers/prebid.js [hunk 5]
```diff
@@ -31,6 +31,34 @@ define('ext.wikia.adEngine.wrappers.prebid', [
...
+					hasVast = bid.vastUrl || bid.vastContent;
+
+				return canUseThisBidder && hasVast;
+			})
+			.reduce(function (previousBid, currentBid) {
+				if (previousBid === null && currentBid.cpm > 0) {
+					return currentBid;
+				}
+
+				if (currentBid !== null && currentBid.cpm > previousBid.cpm) {
```
Comment (WikiaTech) [inline extensions/wikia/AdEngine/js/wrappers/prebid.js]: CRITICAL TypeError can be thrown as "previousBid" might be null or undefined here. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS2259)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue
* ![CRITICAL][CRITICAL] 1 critical

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

**Review by kvas-damian — APPROVED** [review]

**Pull request merged**